### PR TITLE
Implementation of the GET /api/clients endpoint.

### PR DESCRIPTION
**Pull request by jkwatson**

### CODE REVIEW

**Comment by botandrose** [comment]
Taking a look at this now!

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.eclipse.jetty.http.HttpStatus
+import spark.Response
+import javax.inject.Inject
+
+class ClientRoutes @Inject constructor(private val clientDataAccess: ClientDataAccess) {
+
+    val getAll = {
+        val clients = clientDataAccess.allClients()
+        JsonApiResponse(clients.map { it.toJsonApiObject() })
```
Comment (botandrose) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt]: This is interesting. `it` is akin to JavaScript `this` in anonymous functions?

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    val create = { input: JsonApiRequest, response: Response ->
+        val client = Client(input.data.attributes)
```
Comment (botandrose) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt]: Its crazy how jealous I am of simple json destructuring after spending the last day in Haskell. :)

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/Data.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.sleazyweasel.eboshi
+
+import kotlin.properties.getValue
+
+data class Client(val data: Map<String, Any?>) {
+    val id: Int? by data;
+}
+
+data class JsonApiObject(val id: String?, val type: String, val attributes: Map<String, Any?>)
```
Comment (botandrose) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/Data.kt]: Cool, `String?` == built-in optional types?

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/SparkGson.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            handler()
+        }, gsonTransformer)
+    }
+
+    fun port(port: Int) = Spark.port(port)
+
+    fun get(path: String, route: (request: Request, response: Response) -> Any?) = Spark.get(path, route)
+
+}
+
```
Comment (botandrose) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/SparkGson.kt]: I like this decorator approach. Very decoupled and reusable. :+1:

**Comment by botandrose** [comment]
This is very cool. Kotlin definitely appears to be living up to its description as a modernized Java.

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt (hunk 1 above)
Comment (jkwatson) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/ClientRoutes.kt]: yep.

File: kotlin_spark/src/main/java/com/sleazyweasel/eboshi/Data.kt (hunk 3 above)
Comment (jkwatson) [inline kotlin_spark/src/main/java/com/sleazyweasel/eboshi/Data.kt]: yes, indeed

**Pull request merged**